core/base/segmented_range.hpp

```diff
@@ -0,0 +1,361 @@
+     *
+     * @param segment  the index to access. It must be in `[0, num_segments())`.
+     * @return  the segment at this index.
+     */
+    constexpr segment operator[](index_type segment) const
+    {
+        assert(segment >= 0);
+        assert(segment < num_segments());
+        return (*iterator{*this, segment}).segment;
+    }
+
+    /** @return the number of segments in this range. */
+    constexpr index_type num_segments() const { return num_segments_; }
+
+    /** @return iterator pointing to the first segment. */
+    constexpr iterator begin() const { return iterator{*this, 0}; }
+
+    /** @return iterator pointing one past the last segment. */
+    constexpr iterator end() const { return iterator{*this, num_segments()}; }
+
+    /** @return iterator pointing to the first segment. */
+    constexpr const index_type* begin_indices() const { return begins_; }
+
+    /** @return iterator pointing one past the last segment. */
+    constexpr const index_type* end_indices() const { return ends_; }
+
+    /** @return the beginning index of the given segment. */
+    constexpr index_type begin_index(index_type segment) const
+    {
+        assert(segment >= 0);
+        assert(segment < num_segments());
+        return begin_indices()[segment];
+    }
+
+    /** @return the end index of the given segment. */
+    constexpr index_type end_index(index_type segment) const
+    {
+        assert(segment >= 0);
+        assert(segment < num_segments());
+        return end_indices()[segment];
+    }
+
+    /** Compares two ranges for equality. */
+    constexpr friend bool operator==(segmented_range lhs, segmented_range rhs)
+    {
+        return lhs.begin_indices() == rhs.begin_indices() &&
+               lhs.end_indices() == rhs.end_indices() &&
+               lhs.num_segments() == rhs.num_segments();
+    }
+
+    /** Compares two ranges for inequality. */
+    constexpr friend bool operator!=(segmented_range lhs, segmented_range rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+private:
+    const index_type* begins_;
+    const index_type* ends_;
+    index_type num_segments_;
+};
+
+
+/**
+ * Represents a range of indices that is segmented into contiguous segments,
+ * mapped into a value array. Each segment has the shape `[begin, end)`, i.e. it
+ * is a half-open interval and points to corresponding entries of the value
+ * array.
+ *
+ * @tparam IndexType  the type of indices used to represent the segments.
+ * @tparam ValueIterator  the iterator type pointing to the values.
+ */
+template <typename IndexType, typename ValueIterator>
+class segmented_value_range {
+public:
+    using index_type = IndexType;
+    using index_iterator = index_iterator<index_type>;
+    using value_iterator = ValueIterator;
+    using segment = iterator_range<ValueIterator>;
+    using enumerated_range = segmented_value_range<
+        index_type, detail::zip_iterator<index_iterator, value_iterator>>;
```

Comment:
this class will not work on the device, because `zip_iterator` doesn't work on the device. Maybe different zip iterator types will be used depending on the backend?